google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +403,52 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        outstandingReceipts.put(
+            message.getAckId(), new ReceiptCompleteData(outstandingMessage, false));
+      } else {
+        outstandingBatch.add(outstandingMessage);
+      }
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  synchronized void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      ReceiptCompleteData receiptCompleteData = outstandingReceipts.get(ackRequestData.getAckId());
+      OutstandingMessage outstandingMessage = receiptCompleteData.getOutstandingMessage();
+      // Setting to true means that the receipt is complete
+      receiptCompleteData.setReceiptComplete(true);
+      List<OutstandingMessage> completedReceipts = new ArrayList<>();
+
+      for (Iterator<Entry<String, ReceiptCompleteData>> it =
+              outstandingReceipts.entrySet().iterator();
+          it.hasNext(); ) {
+        Map.Entry<String, ReceiptCompleteData> receipt = it.next();
+        String ackId = receipt.getKey();
+        // If receipt is complete then add to completedReceipts to process the batch
+        if (receipt.getValue().getReceiptComplete()) {
+          it.remove();
+          pendingMessages.putIfAbsent(
+              outstandingMessage.receivedMessage.getAckId(), outstandingMessage.ackHandler);
+          completedReceipts.add(receipt.getValue().getOutstandingMessage());
+        } else {
+          break;
+        }
+      }
+      processBatch(completedReceipts);
+    }
+  }
+
+  synchronized void notifyAckFailed(AckRequestData ackRequestData) {
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
```

Comment:
You don't need to check if outstandingReceipts contains the ackId, calling remove will just return null if it's not found.